def.cf

```diff
@@ -4,30 +4,73 @@
 #  - common/global variables and classes here
 #
 ###############################################################################
+body file control
+{
+      inputs => { @(def.extra_inputs) };
+}
 
 bundle common def
 
 {
+  classes:
+      # all extra classes are defined true
+      "$(extra_classes)" expression => "any";
+
+      # all override classes are defined true
+      "$(override_classes)" expression => "any";
+
+      "$(extra_roles[byname][$(roles_byname_keys)])" expression => classmatch("$(roles_byname_keys)");
+      "$(roles_byrole_keys)" expression => classmatch("$(extra_roles[byrole][$(roles_byrole_keys)])");
 
   vars:
 
     any::
+      "defvars" slist => variablesmatching("default:def\..*", "defvar");
+
+      "extra_classes" slist => readstringlist("$(this.promise_dirname)/services/extra_classes.txt", "#.*", "\s", 10k, 100k);
+
+      "augments" data => readjson("$(this.promise_dirname)/services/def.json", 100k);
```

Comment:
It might be nice if def.json were relocatable, also I think it should only be read if it exists.